File: 2021_Samsung_SDS_algorithm_lecture/using_Cpp/15663.cpp
/*
input :
4 2
9 7 9 1

output :
1 7
1 9
7 1
7 9
9 1
9 7
9 9
*/

#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

// 전역 변수 선언
int N, M;
int nums[8];
int ans[8];
int is_used[8];

void recur(int k){
    if (k == M) {
        for (int i = 0; i < M; i++){
            cout << ans[i] << " ";
        }
        cout << '\n';
        return;
    }

    int prev = 0;
    for (int i = 0; i < N; i++){
        if (is_used[i]) continue;
        if (prev == nums[i]) continue;

        ans[k] = nums[i];
        is_used[i] = true;
        prev = nums[i];
        recur(k + 1);
        is_used[i] = false;
    }
}

int main(){
    ios::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
    freopen("input.txt", "r", stdin);

    cin >> N >> M;
    for (int i = 0; i < N; i++){
        cin >> nums[i];
    }
    sort(nums, nums + N);
    recur(0);

    return 0 ;
}

File: 2021_random_set_self_practice/using_python/15663.py
"""
input :
4 2
9 7 9 1

output :
1 7
1 9
7 1
7 9
9 1
9 7
9 9
"""
import sys

sys.stdin = open("input.txt", 'r')
input = sys.stdin.readline

chk = set()


def recur(seq):
    global N, M
    if len(seq) == M:
        if tuple(seq) in chk:
            return
        chk.add(tuple(seq))
        print(' '.join(list(map(str, seq))))
        return

    for i in range(N):
        if is_used[i]: continue

        seq.append(nums[i])
        is_used[i] = True
        recur(seq)
        seq.pop()
        is_used[i] = False


if __name__ == "__main__":
    N, M = map(int, input().rstrip().split())
    nums = list(map(int, input().rstrip().split()))
    nums.sort()
    is_used = [False for _ in range(N)]

    recur([])
